pkg/tsdb/cloudwatch/response_parser.go

```diff
@@ -0,0 +1,155 @@
+			if *message.Code == "MaxMetricsExceeded" {
+				requestExceededMaxLimit = true
+			}
+		}
+
+		for _, r := range mdo.MetricDataResults {
+			if _, ok := mdr[*r.Id]; !ok {
+				mdr[*r.Id] = make(map[string]*cloudwatch.MetricDataResult)
+				mdr[*r.Id][*r.Label] = r
+			} else if _, ok := mdr[*r.Id][*r.Label]; !ok {
+				mdr[*r.Id][*r.Label] = r
+			} else {
+				mdr[*r.Id][*r.Label].Timestamps = append(mdr[*r.Id][*r.Label].Timestamps, r.Timestamps...)
+				mdr[*r.Id][*r.Label].Values = append(mdr[*r.Id][*r.Label].Values, r.Values...)
+			}
+			queries[*r.Id].RequestExceededMaxLimit = requestExceededMaxLimit
+		}
+	}
+
+	cloudWatchResponses := make([]*cloudwatchResponse, 0)
+	for id, lr := range mdr {
+		response := &cloudwatchResponse{}
+		series, err := parseGetMetricDataTimeSeries(lr, queries[id])
+		if err != nil {
+			return cloudWatchResponses, err
+		}
+
+		response.series = series
+		response.SearchExpression = queries[id].SearchExpression
+		response.RefId = queries[id].RefId
+		response.Id = queries[id].Id
+		response.RequestExceededMaxLimit = queries[id].RequestExceededMaxLimit
+
+		cloudWatchResponses = append(cloudWatchResponses, response)
+	}
+
+	return cloudWatchResponses, nil
+}
+
+func parseGetMetricDataTimeSeries(metricDataResults map[string]*cloudwatch.MetricDataResult, query *cloudWatchQuery) (*tsdb.TimeSeriesSlice, error) {
+	result := tsdb.TimeSeriesSlice{}
+	for label, metricDataResult := range metricDataResults {
+		if *metricDataResult.StatusCode != "Complete" {
+			return &result, fmt.Errorf("Part of query failed: %s", *metricDataResult.StatusCode)
+		}
+
+		for _, message := range metricDataResult.Messages {
+			if *message.Code == "ArithmeticError" {
+				return &result, &queryError{fmt.Errorf("ArithmeticError in query %s: %s", query.RefId, *message.Value), query.RefId}
+			}
+		}
+
+		series := tsdb.TimeSeries{
+			Tags:   map[string]string{},
+			Points: make([]tsdb.TimePoint, 0),
+		}
+
+		for key, values := range query.Dimensions {
+			if len(values) == 1 && values[0] != "*" {
+				series.Tags[key] = values[0]
+			} else {
+				for _, value := range values {
+					if value == label || value == "*" {
+						series.Tags[key] = label
+					}
+				}
+			}
+		}
+
+		series.Name = formatAlias(query, query.Stats, series.Tags, label)
+
+		for j, t := range metricDataResult.Timestamps {
+			if j > 0 {
+				expectedTimestamp := metricDataResult.Timestamps[j-1].Add(time.Duration(query.Period) * time.Second)
+				if expectedTimestamp.Before(*t) {
+					series.Points = append(series.Points, tsdb.NewTimePoint(null.FloatFromPtr(nil), float64(expectedTimestamp.Unix()*1000)))
+				}
+			}
+			series.Points = append(series.Points, tsdb.NewTimePoint(null.FloatFrom(*metricDataResult.Values[j]), float64((*t).Unix())*1000))
+		}
+		result = append(result, &series)
+	}
+	return &result, nil
+}
+
+func formatAlias(query *cloudWatchQuery, stat string, dimensions map[string]string, label string) string {
```

Comment:
Could consider adding a few tests of this.

Reply:
Totally agree!